server/backends/migration_cache/migration_cache.go

```diff
@@ -121,46 +129,108 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
+func (mc *MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("not yet implemented")
+func (mc *MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+func (mc *MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+// TODO(Maggie): Add copying logic
+func (mc *MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcBuf []byte
+
+	eg.Go(func() error {
+		srcBuf, srcErr = mc.src.Get(gctx, d)
+		return srcErr
+	})
+
+	// Double read some proportion to guarantee that data is consistent between caches
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.Get(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if doubleRead {
+		if srcErr != dstErr {
+			// Don't log if data not found in dest cache, bc it may not have been copied yet
+			if !status.IsNotFoundError(dstErr) {
+				log.Warningf("Double read of %q failed. src err %s, dest err %s", d, srcErr, dstErr)
+			}
+		}
+	}
+
+	// Return data from source cache
+	return srcBuf, srcErr
+}
+
+func (mc *MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+
+	// Double write data to both caches
+	eg.Go(func() error {
+		srcErr = mc.src.Set(gctx, d, data)
+		return srcErr
+	})
+
+	eg.Go(func() error {
+		dstErr = mc.dest.Set(gctx, d, data)
+		return nil // don't fail if there's an error from this cache
+	})
+
+	if err := eg.Wait(); err != nil {
+		if dstErr == nil {
+			// If error during write to source cache (source of truth), must delete from destination cache
+			deleteErr := mc.dest.Delete(ctx, d)
+			if deleteErr != nil && !status.IsNotFoundError(deleteErr) {
+				log.Warningf("Migration double write err: src write of digest %v failed, but could not delete from dest cache: %s", d, deleteErr)
+			}
+		}
+		return err
+	}
+
+	if dstErr != nil {
```

Comment:
should this be comparing srcErr and dstErr or do we always wanna log if dstErr != nil?

Reply:
srcErr will always be nil here, otherwise we'll exit early on line 228. Modified Get to only check if dstErr == nil too

We could compare srcErr and dstErr before the early exit, but that feels unstable if the new cache returns a slightly different error message. It feels to me like if there's a src error, it doesn't really matter what the dest cache returns. But if the dest cache is erroring a lot, we'd want to know about it